xoinvader/collision.py

```diff
@@ -1,49 +1,188 @@
+        :type col2: :class:`Collider`
+        :rtype: tuple of two :class:`Point`
+        """
+
+        w1 = max(map(len, col1.map))
+        h1 = len(col1.map)
+        tl1 = col1.pos
+        rb1 = tl1 + Point(w1, h1)
+
+        w2 = max(map(len, col2.map))
+        h2 = len(col2.map)
+        tl2 = col2.pos
+        rb2 = tl2 + Point(w2, h2)
+        if (tl1.x >= rb2.x or tl1.y >= rb2.y or
+                rb1.x <= tl2.x or rb1.y <= tl2.y):
+            # Definelty not overlapping
+            return
+        # Now find where exactly overlapping occured
+        tl_ovlp = Point(max(tl1.x, tl2.x), max(tl1.y, tl2.y))
+        rb_ovlp = Point(min(rb1.x, rb2.x), min(rb1.y, rb2.y))
+        # Now find if they actually collided
+        # first, calculate offsets
+        ox1 = abs(tl1.x - tl_ovlp.x)
+        oy1 = abs(tl1.y - tl_ovlp.y)
+        ox2 = abs(tl2.x - tl_ovlp.x)
+        oy2 = abs(tl2.y - tl_ovlp.y)
+        # iterate over overlapping region
+        # and search for collision
+        for i in range(rb_ovlp.x - tl_ovlp.x):
+            for j in range(rb_ovlp.y - tl_ovlp.y):
+                # TODO: check length of current y-level string
+                # it might be not enough to contain i + ox1/2 element
+                if (col1.map[j + oy1][i + ox1] ==
+                        col2.map[j + oy2][i + ox2] == "#"):
+                    return (tl_ovlp, rb_ovlp)
+
+    def _add_collision(self, collider, other_type, callback=None):
+        if collider not in self._colliders:
+            raise ValueError("Attempt to add collision handler for "
+                             "unregistered collider {0}".format(collider))
+        type_pair = TypePair(collider.type, other_type)
+        if type_pair not in self._collisions.keys():
+            self._collisions[type_pair] = [callback]
+        else:
+            self._collisions[type_pair].append(callback)
+
+
+class Collider(object):
+    """Collider component class.
+
+    When added to object, enables it to participate in coliision processing
+    system: i.e. to be able to detect and process collisions between the object
+    and other ones.
+
+    :param str type: name of the collider type. Used in processing possible
+    collisions
+    :param list map: list of strings representing collider geometry. Must be of
+    equal length. "#" character represents actual geometry, all other chars are
+    treated as void space and may be any.
+    :param pos: left top corner of collider map
+    :type pos: :class:`Point`
+    """
+
+    __manager__ = None
+
+    def __init__(self, type, map, pos):
+        if Collider.__manager__ is None:
+            raise ValueError("You can't use Collider objects without "
+                             "ColliderManager. Please create it first.")
+        self._mgr = Collider.__manager__() # manager is weakreaf, thus '()'
+        self._type = type
+        self._pos = pos
+        self._map = map
+        self._mgr._add(self)
```

Comment:
Well, this method is not intended to be called from anywhere but `Collider` class. Thus it's private. And since these both classes are mine, I can do what I want with them.

The point is that public methods are public API, and private methods are okay to call from inside the module.